File: apps/web/src/lib/api/cases.ts
import type { CollectionResponse } from "@/lib/api/projects";

import { apiRequest } from "@/lib/api/client";

export type CaseStatus = string;
export type CaseVersionStatus = string;
export type PromptVersionStatus = string;

export interface PromptVersion {
  id: string;
  prompt_id: string;
  version_number: number;
  status: PromptVersionStatus;
  content: string;
  predecessor_version_id: string | null;
  created_at: string;
}

export interface CaseVersion {
  id: string;
  case_id: string;
  version_number: number;
  status: CaseVersionStatus;
  description: string;
  repository_url: string;
  commit_sha: string;
  subdirectory: string | null;
  expected_checks: string[];
  applicable_grader_ids: string[];
  prompt_version_id: string;
  predecessor_version_id: string | null;
  created_at: string;
}

export interface Case {
  id: string;
  project_id: string;
  prompt_id: string;
  name: string;
  description: string;
  status: CaseStatus;
  created_at: string;
  active_version_id: string | null;
  active_prompt_version_id: string | null;
  versions: CaseVersion[];
  prompt_versions: PromptVersion[];
}

export interface ListCasesParams {
  project_id: string;
  cursor?: string;
  limit?: number;
  sort?: string;
  status?: string;
  q?: string;
}

export interface CreateCaseInput {
  project_id: string;
  name: string;
  description?: string;
}

export interface CreatePromptDraftInput {
  content: string;
}

export interface CreateCaseDraftInput {
  description: string;
  repository_url: string;
  commit_sha: string;
  expected_checks?: string[];
  applicable_grader_ids?: string[];
  prompt_version_id: string;
  subdirectory?: string | null;
}

function buildQuery(params: ListCasesParams): string {
  const search = new URLSearchParams();
  search.set("project_id", params.project_id);
  if (params.cursor) search.set("cursor", params.cursor);
  if (params.limit !== undefined) search.set("limit", String(params.limit));
  if (params.sort) search.set("sort", params.sort);
  if (params.status) search.set("status", params.status);
  if (params.q) search.set("q", params.q);
  return `?${search.toString()}`;
}

export async function listCases(
  token: string,
  params: ListCasesParams,
): Promise<CollectionResponse<Case>> {
  return apiRequest<CollectionResponse<Case>>(`/v1/cases${buildQuery(params)}`, {
    method: "GET",
    token,
  });
}

export async function getCase(token: string, caseId: string): Promise<Case> {
  return apiRequest<Case>(`/v1/cases/${encodeURIComponent(caseId)}`, {
    method: "GET",
    token,
  });
}

export async function createCase(
  token: string,
  input: CreateCaseInput,
  idempotencyKey?: string,
): Promise<Case> {
  const headers: Record<string, string> = {};
  if (idempotencyKey) {
    headers["Idempotency-Key"] = idempotencyKey;
  }
  return apiRequest<Case>("/v1/cases", {
    method: "POST",
    token,
    headers,
    body: {
      project_id: input.project_id,
      name: input.name,
      description: input.description ?? "",
    },
  });
}

export async function createPromptDraftVersion(
  token: string,
  caseId: string,
  input: CreatePromptDraftInput,
): Promise<PromptVersion> {
  return apiRequest<PromptVersion>(`/v1/cases/${encodeURIComponent(caseId)}/prompts/versions`, {
    method: "POST",
    token,
    body: { content: input.content },
  });
}

export async function publishPromptVersion(
  token: string,
  caseId: string,
  versionId: string,
): Promise<PromptVersion> {
  return apiRequest<PromptVersion>(
    `/v1/cases/${encodeURIComponent(caseId)}/prompts/versions/${encodeURIComponent(versionId)}/publish`,
    { method: "POST", token },
  );
}

export async function createCaseDraftVersion(
  token: string,
  caseId: string,
  input: CreateCaseDraftInput,
): Promise<CaseVersion> {
  return apiRequest<CaseVersion>(`/v1/cases/${encodeURIComponent(caseId)}/versions`, {
    method: "POST",
    token,
    body: {
      description: input.description,
      repository_url: input.repository_url,
      commit_sha: input.commit_sha,
      expected_checks: input.expected_checks ?? [],
      applicable_grader_ids: input.applicable_grader_ids ?? [],
      prompt_version_id: input.prompt_version_id,
      subdirectory: input.subdirectory ?? null,
    },
  });
}

export async function publishCaseVersion(
  token: string,
  caseId: string,
  versionId: string,
): Promise<CaseVersion> {
  return apiRequest<CaseVersion>(
    `/v1/cases/${encodeURIComponent(caseId)}/versions/${encodeURIComponent(versionId)}/publish`,
    { method: "POST", token },
  );
}

export async function deprecateCase(token: string, caseId: string): Promise<Case> {
  return apiRequest<Case>(`/v1/cases/${encodeURIComponent(caseId)}/deprecate`, {
    method: "POST",
    token,
  });
}

File: apps/api/src/agent_eval_api/routers/v1/cases.py
"""Case endpoints — invoke Application use cases only."""

from __future__ import annotations

from typing import Annotated

from agent_eval_application.commands.case import (
    CreateCaseCommand,
    CreateCaseDraftVersionCommand,
    DeprecateCaseCommand,
    PublishCaseVersionCommand,
)
from agent_eval_application.queries.queries import GetCaseQuery, ListCasesByProjectQuery
from fastapi import APIRouter, Depends, Header, Query, status

from agent_eval_api.dependencies import ActorDep, ServicesDep
from agent_eval_api.pagination import ListParams
from agent_eval_api.schemas.case import (
    CaseResponse,
    CaseVersionResponse,
    CreateCaseDraftVersionRequest,
    CreateCaseRequest,
)
from agent_eval_api.schemas.common import CollectionResponse

router = APIRouter(prefix="/v1/cases", tags=["cases"])


@router.post("", response_model=CaseResponse, status_code=status.HTTP_201_CREATED)
def create_case(
    body: CreateCaseRequest,
    actor: ActorDep,
    services: ServicesDep,
    idempotency_key: str | None = Header(default=None, alias="Idempotency-Key"),
) -> CaseResponse:
    dto = services.create_case.execute(
        CreateCaseCommand(
            actor=actor,
            project_id=body.project_id,
            name=body.name,
            description=body.description,
            category=body.category,
            difficulty=body.difficulty,
            language=body.language,
            tags=tuple(body.tags),
            idempotency_key=idempotency_key,
        )
    )
    return CaseResponse.from_dto(dto)


@router.get("", response_model=CollectionResponse[CaseResponse])
def list_cases(
    actor: ActorDep,
    services: ServicesDep,
    params: Annotated[ListParams, Depends()],
    project_id: str = Query(min_length=1),
) -> CollectionResponse[CaseResponse]:
    items = services.list_cases_by_project.execute(
        ListCasesByProjectQuery(actor=actor, project_id=project_id)
    )
    responses = [CaseResponse.from_dto(c) for c in items]
    return params.apply(responses)


@router.get("/{case_id}", response_model=CaseResponse)
def get_case(
    case_id: str,
    actor: ActorDep,
    services: ServicesDep,
) -> CaseResponse:
    dto = services.get_case.execute(GetCaseQuery(actor=actor, case_id=case_id))
    return CaseResponse.from_dto(dto)


@router.post(
    "/{case_id}/versions",
    response_model=CaseVersionResponse,
    status_code=status.HTTP_201_CREATED,
)
def create_case_draft_version(
    case_id: str,
    body: CreateCaseDraftVersionRequest,
    actor: ActorDep,
    services: ServicesDep,
) -> CaseVersionResponse:
    dto = services.create_case_draft_version.execute(
        CreateCaseDraftVersionCommand(
            actor=actor,
            case_id=case_id,
            description=body.description,
            repository_url=body.repository_url,
            commit_sha=body.commit_sha,
            expected_checks=tuple(body.expected_checks),
            applicable_grader_ids=tuple(body.applicable_grader_ids),
            prompt_version_id=body.prompt_version_id,
            subdirectory=body.subdirectory,
        )
    )
    return CaseVersionResponse.from_dto(dto)


@router.post(
    "/{case_id}/versions/{version_id}/publish",
    response_model=CaseVersionResponse,
)
def publish_case_version(
    case_id: str,
    version_id: str,
    actor: ActorDep,
    services: ServicesDep,
) -> CaseVersionResponse:
    dto = services.publish_case_version.execute(
        PublishCaseVersionCommand(actor=actor, case_id=case_id, version_id=version_id)
    )
    return CaseVersionResponse.from_dto(dto)


@router.post("/{case_id}/deprecate", response_model=CaseResponse)
def deprecate_case(
    case_id: str,
    actor: ActorDep,
    services: ServicesDep,
) -> CaseResponse:
    dto = services.deprecate_case.execute(
        DeprecateCaseCommand(actor=actor, case_id=case_id)
    )
    return CaseResponse.from_dto(dto)
